CDCgov/prime-simplereport
# Add axe testing for cypress

# TESTING/DEVOPS PULL REQUEST

## Related Issue #4530

- Adds axe accessibility testing to existing cypress e2e workflow tests

## Changes Proposed

- Added npm `cypress-axe` and `axe-core` dependencies to dev
- Added `cy.checkA11y()` for each page/screen/modal visited in the e2e flow for each spec
- **Note that no new workflows were added, so if there wasn't an existing e2e test for a page, that page is not getting accessibility tested.** Additional workflow testing with accessibility can be added in the future. This PR is intended to just build on what already exists.
- For existing issues, rather than comment out the `checkA11y()`, I've added an `exclude` and/or `rules` block to the check, which allows the excluded areas to fail.  This way, we can ignore existing problems, but still test the rest of the page to catch regressions. These will need to be taken out once the underlying issues have been resolved.

## Additional Information

For future reference:
- `checkA11y()` can produce strange results if called too early (e.g. the page hasn't loaded).
- a11y tests are most reliable directly after an assertion like `.get` or `.contains` that does not do an action (like a `.click`), just checks that what we expect was loaded - again this avoids strange issues due to a page not loading in time.

Testing for admin pages - these are rife with errors
- I've added tests for these and marked them `// failing a11y test - admin`. Because these are not public facing they are low priority, we won't create a ticket for these at this time 

I've left `// failing a11y test` comments to indicate that a test should be enabled, but will fail without an accessibility fix on the page or component in question. When the following tickets are worked, it should be easy to identify what needs to be fixed by looking for these comments.

Creating tickets for:
 - Toast, which causes errors on multiple pages.
 - Address verification Modal, which appears on multiple pages. Once this is resolved, the exclusions/rules for these tests across specs can be removed
 - Tickets for each other spec that is failing due to other isolated a11y issues - these are isolated to the pages being tested and can be worked independently

## Testing

- Try running the e2es on locally and check out the last run in the pipeline PR check
- Confirm that I've covered pages visited in the e2e workflows - at this time I don't want to add any new workflows, just test what we are currently visiting.


### Documentation
- I'll be adding a wiki page for a11y testing - [New wiki page is here](https://github.com/CDCgov/prime-simplereport/wiki/Accessibility-Auditing)
